zerver/lib/api.py

```diff
@@ -0,0 +1,334 @@
+        # Set metadata to a default early to avoid addition None-checks.
+        if metadata is None:
+            metadata = Metadata()
+
+        # Metadata defines a validator making sure that argument_type_is_body is
+        # incompatible with whence.
+        if metadata.argument_type_is_body:
+            request_var_name = "request"
+        else:
+            request_var_name = metadata.whence if metadata.whence is not None else param_name
+
+        process_parameters.append(
+            FuncParam(
+                param_name=param_name,
+                param_type=param_type,
+                request_var_name=request_var_name,
+                default=param_default,
+                metadata=metadata,
+            )
+        )
+
+    return ViewFuncInfo(
+        view_func_full_name=view_func_full_name,
+        parameters=process_parameters,
+        no_nesting=True,
+    )
+
+
+def parse_request(
+    request: HttpRequest,
+    view_func_info: ViewFuncInfo,
+    args: Sequence[object],
+    kwargs: Mapping[str, object],
+) -> CallInfo:
+    processed_request_var_names = set()
+    kwargs_to_update = {}
+    errors = []
+    for parameter in view_func_info.parameters:
+        if parameter.metadata.path_only:
+            # For path_only parameters, they should already have been passed via
+            # the URL, so there's no need for us to do anything.
+            #
+            # TODO: Run validators for path_only parameters for NewType.
+            assert (
+                parameter.param_name in kwargs
+            ), f"Path-only variable {parameter.param_name} should be passed already"
+        if parameter.param_name in kwargs:
+            # Skip parameters that are already supplied by the caller.
+            continue
+
+        # Extract the value to parse from the request.
+        value_to_parse: str
+        request_var_name = parameter.request_var_name
+        if parameter.metadata.argument_type_is_body:
+            try:
+                value_to_parse = request.body.decode(request.encoding or "utf-8")
+            except UnicodeDecodeError:
+                raise JsonableError(_("Malformed payload"))
+        elif request_var_name in request.POST:
+            value_to_parse = request.POST[request_var_name]
+            processed_request_var_names.add(request_var_name)
+        elif request_var_name in request.GET:
+            value_to_parse = request.GET[request_var_name]
+            processed_request_var_names.add(request_var_name)
+        else:
+            if parameter.default is NotSpecified:
+                raise RequestVariableMissingError(request_var_name)
+            # By skipping here, we leave it to Python to use the default value
+            # of this parameter, because we cannot find the request variable in
+            # the request.
+            # This is covered by test_json, but coverage.py
+            # fails to recognize this for some reason.
+            continue  # nocoverage
```

Comment:
Python does optimizations with `continue` that confuse `coverage.py`.  Whenever there's a non-contorted way to avoid `continue` in Python, which usually entails extracting some kind of helper function or re-arranging conditionals, I try to go that route.  If that's impractical here, then I think this comment is sufficient.  But I'd give it a quick look to see if there's a better way to phrase this code.  Long if/elif/elif/elif blocks are also kind of a code smell, but, again, sometimes unavoidable.

Reply:
Added a `fallback_to_defaults` dict before this line to make coverage work.